### PR TITLE
avm2: Implement setStencilReferenceValue and setStencilActions on flash.display3D.Context3D

### PR DESCRIPTION
**Pull request by SuchAFuriousDeath**
Implements #14000 (Tested and it works properly now).

### CODE REVIEW

**Comment by kjarosh** [comment]
We'll need some tests for it, probably a visual test to check that it renders properly, and a test to verify the API (it can be one test if that's easier, but usually we focus on different things there).

File: render/src/backend.rs [hunk 1]
```diff
@@ -296,6 +296,22 @@ pub enum Context3DTriangleFace {
     FrontAndBack,
 }
 
+impl Context3DTriangleFace {
+    pub fn from_wstr(s: &WStr) -> Option<Self> {
+        if s == b"none" {
+            Some(Context3DTriangleFace::None)
+        } else if s == b"back" {
+            Some(Context3DTriangleFace::Back)
+        } else if s == b"Front" {
```
Comment (kjarosh) [inline render/src/backend.rs]: Edge cases like this one should be tested too.

**Comment by SuchAFuriousDeath** [comment]
I'll put this in draft until I get the tests done.

File: render/src/backend.rs [hunk 2]
```diff
@@ -429,6 +429,76 @@ impl FromWStr for Context3DCompareMode {
     }
 }
 
+#[cfg(feature = "wgpu")]
```
Comment (Lord-McSweeney) [inline render/src/backend.rs]: Could this be moved to the `wgpu` part of `render` to avoid having to feature-gate it?

File: render/src/backend.rs [hunk 3]
```diff
@@ -429,6 +429,76 @@ impl FromWStr for Context3DCompareMode {
...
+            Ok(Context3DStencilAction::Set)
+        } else if s == b"zero" {
+            Ok(Context3DStencilAction::Zero)
+        } else {
+            Err(())
+        }
+    }
+}
+
+#[cfg(feature = "wgpu")]
```
Comment (Lord-McSweeney) [inline render/src/backend.rs]: Here too

File: render/src/backend.rs (hunk 2 above)
Comment (SuchAFuriousDeath) [reply render/src/backend.rs]: Actually it couldn't, because of the orphan rule. In render, the Contexd3D nor the wgpu types are local to that crate :(

Comment (kjarosh) [reply render/src/backend.rs]: In that case you can use a named method instead of the `From` trait. I don't like the fact that it's implemented behind a feature too especially that we have a dedicated crate for its implementation.

Comment (SuchAFuriousDeath) [reply render/src/backend.rs]: I guess I'll create an IntoWgpu trait, otherwise it's annoying to name these functions :sweat_smile:

File: core/src/avm2/globals/flash/display3D/context_3d.rs [hunk 4]
```diff
@@ -623,6 +623,53 @@ pub fn set_render_to_texture<'gc>(
...
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let this = this.as_object().unwrap();
+
+    if let Some(context) = this.as_context_3d() {
+        let triangle_face = args.get_string(activation, 0);
+
+        let triangle_face = triangle_face
+            .parse()
+            .inspect_err(|_| tracing::error!("Unknown triangle_face {:?}", triangle_face))
```
Comment (kjarosh) [inline core/src/avm2/globals/flash/display3D/context_3d.rs]: It's not really an error—we either (1) don't know what happens in this case (a warning stub? a TODO?) or (2) we handle the case properly (no logs).

If you're not sure whether we handle that properly go with (1), if you're sure go with (2). Of course I would recommend writing a test to check this edge case.

File: core/src/avm2/globals/flash/display3D/context_3d.rs [hunk 5]
```diff
@@ -623,6 +623,53 @@ pub fn set_render_to_texture<'gc>(
...
+            .unwrap_or(Context3DTriangleFace::None);
+
+        let compare_mode = args
+            .get_string_non_null(activation, 1, "compareMode")?
+            .parse()
+            .map_err(|_| make_error_2008(activation, "compareMode"))?;
```
Comment (kjarosh) [inline core/src/avm2/globals/flash/display3D/context_3d.rs:693]: Can we add tests for these exceptions? It can be a separate, non-image test.

File: core/src/avm2/globals/flash/display3D/context_3d.rs (hunk 4 above)
Comment (kjarosh) [reply core/src/avm2/globals/flash/display3D/context_3d.rs]: According to docs, we should throw an error like below, but should be easy to write a simple test for it.

File: render/wgpu/src/context3d/current_pipeline.rs [hunk 6]
```diff
@@ -122,6 +165,15 @@ impl CurrentPipeline {
 
             depth_mask: true,
             pass_compare_mode: wgpu::CompareFunction::LessEqual,
+
+            stencil: wgpu::StencilState {
+                front: wgpu::StencilFaceState::IGNORE,
+                back: wgpu::StencilFaceState::IGNORE,
+                read_mask: !0,
```
Comment (kjarosh) [inline render/wgpu/src/context3d/current_pipeline.rs]: Better to be explicit: use `u32::MAX` instead of `!0`, you don't have to worry about the type (size, signedness) and you don't have to wonder what that expression means (because it's not commonly used).

File: render/wgpu/src/context3d/current_pipeline.rs [hunk 7]
```diff
@@ -64,6 +104,9 @@ pub struct CurrentPipeline {
     depth_mask: bool,
     pass_compare_mode: wgpu::CompareFunction,
 
+    stencil: wgpu::StencilState,
+    stencil_ref_value: u32,
```
Comment (kjarosh) [inline render/wgpu/src/context3d/current_pipeline.rs:110]: Can you add a doc and explain what that is?

File: render/wgpu/src/context3d/current_pipeline.rs (hunk 6 above)
Comment (kjarosh) [reply render/wgpu/src/context3d/current_pipeline.rs]: Actually now thinking about this, it's a mask so all we care about is that all bits are set. 🤔 What do you think about this?

**Comment by kjarosh** [comment]
Code LGTM, I think we should cover edge cases in tests, currently only the happy path is covered.

**Comment by SuchAFuriousDeath** [comment]
Edge cases with the string parsing (working on that)? Or something more than that?

File: render/wgpu/src/context3d/current_pipeline.rs (hunk 6 above)
Comment (SuchAFuriousDeath) [reply render/wgpu/src/context3d/current_pipeline.rs]: I'm fine with either.

File: core/src/avm2/globals/flash/display3D/context_3d.rs (hunk 5 above)
Comment (SuchAFuriousDeath) [reply core/src/avm2/globals/flash/display3D/context_3d.rs:693]: Done

File: core/src/avm2/globals/flash/display3D/context_3d.rs (hunk 4 above)
Comment (SuchAFuriousDeath) [reply core/src/avm2/globals/flash/display3D/context_3d.rs]: I changed it to get_string_non_null like the other ones, tests included.

**Comment by kjarosh** [comment]
There are some edge cases in `setStencilReferenceValue`, but I don't think they are testable without visuals? If you feel like writing a test for that, great, but I don't think it's strictly necessary, a TODO is fine.

**Comment by SuchAFuriousDeath** [comment]
> There are some edge cases in `setStencilReferenceValue`, but I don't think they are testable without visuals? If you feel like writing a test for that, great, but I don't think it's strictly necessary, a TODO is fine.

https://github.com/ruffle-rs/ruffle/pull/20875/changes#diff-46759425764463d9c159e31f66487e738e5a3917b0ab94e562795703a371aa2fR683

Hopefully this is what you meant.

I think this PR already does enough :sweat_smile: Maybe I'll do it some other time, if I feel like it.

**Comment by SuchAFuriousDeath** [comment]
With regards to the string parsing: Everything is case-sensitive and uses get_string_non_null, except Context3DVertexBufferFormat, which uses get_string.

**Review by Lord-McSweeney — APPROVED** [review]
I'm not very experienced with the wgpu code, but the AS3 part lgtm

**Review by kjarosh — APPROVED** [review]
